src/rust/engine/src/rule_graph.rs

```diff
@@ -837,8 +820,10 @@ fn rhs_for_select(tasks: &Tasks, subject_type: TypeId, select: &Select) -> Entri
   if externs::satisfied_by_type(&select.product, &subject_type) {
     // NB a matching subject is always picked first
     vec![Entry::new_subject_is_product(subject_type)]
+  } else if let Some(&(ref key, _)) = tasks.gen_singleton(&select.product) {
```

Comment:
I didn't realize til now, but this needs to have the new intrinsic, ie snapshot / filecontents, dispatch here too to accurately model how the engine works. I *think* it's ok for now without.